### PR TITLE
network: dhcp-server: introduce UplinkInterface= setting

### PR DESCRIPTION
**Pull request by yuwata**
Replaces #14888 and #14907.

### CODE REVIEW

**Comment by yuwata** [comment]
Rebased. PTAL.

**Comment by bluca** [comment]
Tests on CentOS look unhappy, and it seems related:

```
14:38:43 ======================================================================
14:38:43 ERROR: test_wait_online_ipv6 (__main__.NetworkdNetworkTests)
14:38:43 ----------------------------------------------------------------------
14:38:43 Traceback (most recent call last):
14:38:43   File "./test/test-network/systemd-networkd-tests.py", line 3146, in test_wait_online_ipv6
14:38:43     self.wait_online(['veth99:routable'], ipv6=True)
14:38:43   File "./test/test-network/systemd-networkd-tests.py", line 589, in wait_online
14:38:43     check_output(*args, env=env)
14:38:43   File "./test/test-network/systemd-networkd-tests.py", line 49, in check_output
14:38:43     return subprocess.check_output(command, universal_newlines=True, **kwargs).rstrip()
14:38:43   File "/usr/lib64/python3.6/subprocess.py", line 356, in check_output
14:38:43     **kwargs).stdout
14:38:43   File "/usr/lib64/python3.6/subprocess.py", line 438, in run
14:38:43     output=stdout, stderr=stderr)
14:38:43 subprocess.CalledProcessError: Command '['/usr/lib/systemd/systemd-networkd-wait-online', '--timeout=20s', '--interface=veth99:routable', '--ipv6']' returned non-zero exit status 1.
14:38:43 
14:38:43 ======================================================================
14:38:43 ERROR: test_ipv6_prefix_delegation (__main__.NetworkdRATests)
14:38:43 ----------------------------------------------------------------------
14:38:43 Traceback (most recent call last):
14:38:43   File "./test/test-network/systemd-networkd-tests.py", line 3623, in test_ipv6_prefix_delegation
14:38:43     self.wait_online(['veth99:routable', 'veth-peer:degraded'])
14:38:43   File "./test/test-network/systemd-networkd-tests.py", line 589, in wait_online
14:38:43     check_output(*args, env=env)
14:38:43   File "./test/test-network/systemd-networkd-tests.py", line 49, in check_output
14:38:43     return subprocess.check_output(command, universal_newlines=True, **kwargs).rstrip()
14:38:43   File "/usr/lib64/python3.6/subprocess.py", line 356, in check_output
14:38:43     **kwargs).stdout
14:38:43   File "/usr/lib64/python3.6/subprocess.py", line 438, in run
14:38:43     output=stdout, stderr=stderr)
14:38:43 subprocess.CalledProcessError: Command '['/usr/lib/systemd/systemd-networkd-wait-online', '--timeout=20s', '--interface=veth99:routable', '--interface=veth-peer:degraded']' returned non-zero exit status 1.
14:38:43 
14:38:43 ======================================================================
14:38:43 ERROR: test_ipv6_token_prefixstable (__main__.NetworkdRATests)
14:38:43 ----------------------------------------------------------------------
14:38:43 Traceback (most recent call last):
14:38:43   File "./test/test-network/systemd-networkd-tests.py", line 3653, in test_ipv6_token_prefixstable
14:38:43     self.wait_online(['veth99:routable', 'veth-peer:degraded'])
14:38:43   File "./test/test-network/systemd-networkd-tests.py", line 589, in wait_online
14:38:43     check_output(*args, env=env)
14:38:43   File "./test/test-network/systemd-networkd-tests.py", line 49, in check_output
14:38:43     return subprocess.check_output(command, universal_newlines=True, **kwargs).rstrip()
14:38:43   File "/usr/lib64/python3.6/subprocess.py", line 356, in check_output
14:38:43     **kwargs).stdout
14:38:43   File "/usr/lib64/python3.6/subprocess.py", line 438, in run
14:38:43     output=stdout, stderr=stderr)
14:38:43 subprocess.CalledProcessError: Command '['/usr/lib/systemd/systemd-networkd-wait-online', '--timeout=20s', '--interface=veth99:routable', '--interface=veth-peer:degraded']' returned non-zero exit status 1.
14:38:43 
14:38:43 ======================================================================
14:38:43 ERROR: test_ipv6_token_prefixstable_without_address (__main__.NetworkdRATests)
14:38:43 ----------------------------------------------------------------------
14:38:43 Traceback (most recent call last):
14:38:43   File "./test/test-network/systemd-networkd-tests.py", line 3663, in test_ipv6_token_prefixstable_without_address
14:38:43     self.wait_online(['veth99:routable', 'veth-peer:degraded'])
14:38:43   File "./test/test-network/systemd-networkd-tests.py", line 589, in wait_online
14:38:43     check_output(*args, env=env)
14:38:43   File "./test/test-network/systemd-networkd-tests.py", line 49, in check_output
14:38:43     return subprocess.check_output(command, universal_newlines=True, **kwargs).rstrip()
14:38:43   File "/usr/lib64/python3.6/subprocess.py", line 356, in check_output
14:38:43     **kwargs).stdout
14:38:43   File "/usr/lib64/python3.6/subprocess.py", line 438, in run
14:38:43     output=stdout, stderr=stderr)
14:38:43 subprocess.CalledProcessError: Command '['/usr/lib/systemd/systemd-networkd-wait-online', '--timeout=20s', '--interface=veth99:routable', '--interface=veth-peer:degraded']' returned non-zero exit status 1.
14:38:43 
14:38:43 ======================================================================
14:38:43 ERROR: test_ipv6_token_static (__main__.NetworkdRATests)
14:38:43 ----------------------------------------------------------------------
14:38:43 Traceback (most recent call last):
14:38:43   File "./test/test-network/systemd-networkd-tests.py", line 3641, in test_ipv6_token_static
14:38:43     self.wait_online(['veth99:routable', 'veth-peer:degraded'])
14:38:43   File "./test/test-network/systemd-networkd-tests.py", line 589, in wait_online
14:38:43     check_output(*args, env=env)
14:38:43   File "./test/test-network/systemd-networkd-tests.py", line 49, in check_output
14:38:43     return subprocess.check_output(command, universal_newlines=True, **kwargs).rstrip()
14:38:43   File "/usr/lib64/python3.6/subprocess.py", line 356, in check_output
14:38:43     **kwargs).stdout
14:38:43   File "/usr/lib64/python3.6/subprocess.py", line 438, in run
14:38:43     output=stdout, stderr=stderr)
14:38:43 subprocess.CalledProcessError: Command '['/usr/lib/systemd/systemd-networkd-wait-online', '--timeout=20s', '--interface=veth99:routable', '--interface=veth-peer:degraded']' returned non-zero exit status 1.
14:38:43 
14:38:43 ======================================================================
14:38:43 FAIL: test_ipv6_route_prefix (__main__.NetworkdIPv6PrefixTests)
14:38:43 ----------------------------------------------------------------------
14:38:43 Traceback (most recent call last):
14:38:43   File "./test/test-network/systemd-networkd-tests.py", line 4612, in test_ipv6_route_prefix
14:38:43     self.assertIn('inet6 2001:db8:0:1:', output)
14:38:43 AssertionError: 'inet6 2001:db8:0:1:' not found in '176: veth-peer@veth99: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000\n    link/ether 12:34:56:78:9a:bd brd ff:ff:ff:ff:ff:ff\n    inet 192.168.5.1/24 brd 192.168.5.255 scope global veth-peer\n       valid_lft forever preferred_lft forever\n    inet 192.168.5.2/24 brd 192.168.5.255 scope global secondary veth-peer\n       valid_lft forever preferred_lft forever\n    inet6 fe80::1034:56ff:fe78:9abd/64 scope link \n       valid_lft forever preferred_lft forever'
14:38:43 
14:38:43 ======================================================================
14:38:43 FAIL: test_ipv6_route_prefix_deny_list (__main__.NetworkdIPv6PrefixTests)
14:38:43 ----------------------------------------------------------------------
14:38:43 Traceback (most recent call last):
14:38:43   File "./test/test-network/systemd-networkd-tests.py", line 4635, in test_ipv6_route_prefix_deny_list
14:38:43     self.assertIn('inet6 2001:db8:0:1:', output)
14:38:43 AssertionError: 'inet6 2001:db8:0:1:' not found in '178: veth-peer@veth99: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000\n    link/ether 12:34:56:78:9a:bd brd ff:ff:ff:ff:ff:ff\n    inet 192.168.5.1/24 brd 192.168.5.255 scope global veth-peer\n       valid_lft forever preferred_lft forever\n    inet 192.168.5.2/24 brd 192.168.5.255 scope global secondary veth-peer\n       valid_lft forever preferred_lft forever\n    inet6 fe80::1034:56ff:fe78:9abd/64 scope link \n       valid_lft forever preferred_lft forever'
```

**Comment by yuwata** [comment]
Ouch. I made a mistake on rebase. I hope it is now fixed. PTAL.

File: src/network/networkd-dhcp-server.c [hunk 1]
```diff
@@ -584,3 +660,57 @@ int config_parse_dhcp_server_address(
...
+                           "Invalid interface name in %s=, ignoring assignment: %s", lvalue, rvalue);
+                return 0;
+        }
+
+        name = strdup(rvalue);
+        if (!name)
+                return log_oom();
+
+        network->dhcp_server_uplink_index = 0;
+        network->dhcp_server_uplink_name = name;
```
Comment (poettering) [inline src/network/networkd-dhcp-server.c:714]: as mentioned in the other PR, this is a memory leak if people use UplinkInterface= twice in the same network file.

i.e. use free_and_replace() or free_and_strdup()

**Comment by poettering** [comment]
looks good to merge, except for the leak thing

**Comment by yuwata** [comment]
This is included in #19639, which is already reviewed and updated. Closing.

**Pull request closed without merging**